### DISCUSSION TITLE
Is it possible for Eglot to not clear all diagnostics after every edit?

### DISCUSSION THREAD

**Discussion by andrewbanchich** [joaotavora/eglot#1390] [General]
Using Rust, I noticed if I have e.g. several error diagnostics from Flymake for types I haven't imported, I need to:

1. Use code actions to import one
2. Save
3. All diagnostics are cleared
4. Wait for them to come back
5. Repeat

This can make things pretty slow, depending on the project and if you're using Tramp. Is there any way Eglot could do some of this asynchronously so it can preserve all the Flymake errors and just update them in the background while I'm editing?

**Comment by joaotavora** [comment]
Eglot does not clear all diagnostics after every edit.  If the diagnostic annotations are cleared it's probably because the server sent a `publishDiagnostics` notification saying to do so.

Maybe you should share an excerpt of the Eglot events buffer, hopefully collected in a small enough Rust toy project.  A description of this project and exactly what you do is also welcome, much like requested here: https://joaotavora.github.io/eglot/#Troubleshooting-Eglot

But if you can't do that, just show the Eglot events buffer. around the time of steps 1-4

**Reply by andrewbanchich** [reply]
I found what I'm pretty sure is the root cause of the slowness (not related to this). Thanks for the help!
